### PR TITLE
hv: vlapic: kick targeted vCPU off if interrupt trigger mode has changed

### PR DESCRIPTION
**Pull request by lifeix**
In APICv advanced mode, an targeted vCPU, running in non-root mode, may get outdated
TMR and EOI exit bitmap if another vCPU sends an interrupt to it if the trigger mode
of this interrupt has changed.
This patch try to kick vCPU off to let it get the latest TMR and EOI exit bitmap when
it enters non-root mode again. Then fill the interrupt to PIR.

Tracked-On: #4200
Signed-off-by: Li Fei1 <fei1.li@intel.com>

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
4 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:part_config filename:/hypervisor/arch/x86/guest/vlapic.c func:vlapic_is_tmr_changed offset:508<br> reason:Pointer param should be declared pointer to const. : vlapic<br><2> config:sharing_config filename:/hypervisor/arch/x86/guest/vlapic.c func:vlapic_is_tmr_changed offset:508<br> reason:Pointer param should be declared pointer to const. : vlapic<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vlapic.c func:vlapic_is_tmr_changed offset:511<br> reason:Pointer not checked for null before use. : tmrptr<br><4> config:sharing_config filename:/hypervisor/arch/x86/guest/vlapic.c func:vlapic_is_tmr_changed offset:511<br> reason:Pointer not checked for null before use. : tmrptr<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: hypervisor/arch/x86/guest/vlapic.c [hunk 1]
```diff
@@ -552,19 +555,21 @@ static void apicv_basic_accept_intr(struct acrn_vlapic *vlapic, uint32_t vector,
...
 
 static void apicv_advanced_accept_intr(struct acrn_vlapic *vlapic, uint32_t vector, bool level)
 {
-	int32_t pending_intr = apicv_set_intr_ready(vlapic, vector);
-
-	vlapic_set_tmr(vlapic, vector, level);
+	if (vlapic_is_tmr_changed(vlapic, vector, level)) {
+		/* kick vCPU off to update TMR and EOI exit bitmap if interrupt trigger mode has changed */
+		kick_vcpu(vlapic->vcpu);
+		vlapic_set_tmr(vlapic, vector, level);
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vlapic.c]: set_tmr first or kick_vcpu first?

**Comment by HuiHuangShi** [comment]
2 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:part_config filename:/hypervisor/arch/x86/guest/vlapic.c func:vlapic_is_tmr_changed offset:511<br> reason:Pointer not checked for null before use. : tmrptr<br><2> config:sharing_config filename:/hypervisor/arch/x86/guest/vlapic.c func:vlapic_is_tmr_changed offset:511<br> reason:Pointer not checked for null before use. : tmrptr<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: hypervisor/arch/x86/guest/vlapic.c (hunk 1 above)
Comment (lifeix) [reply hypervisor/arch/x86/guest/vlapic.c]: On second thought, The new TMR and EOI exit bitmap are only used for the new interrupts. It should be OK to kick vCPU off after TMR and EOI exit bitmap has updated and before new interrupts are updated to PIR.

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Pull request merged**